windows/hid.c

```diff
@@ -223,6 +229,9 @@ static void free_hid_device(hid_device *dev)
 	free(dev->feature_buf);
 	free(dev->read_buf);
 	hid_free_enumeration(dev->device_info);
+	DeleteCriticalSection(&dev->cs);
+	DeleteCriticalSection(&dev->cs_read);
+	DeleteCriticalSection(&dev->cs_write);
 	free(dev);
```

Comment:
this is what I mentioned by "not possible to implement a thread-safe `hid_close`. Imagine a scenario:
1) One thread enters a `hid_close` and enter the critical section;
2) Another thread enters `hid_read` and waits on Critical section;
3) First thread destroys a critical section here (is it even allowed w/o first "leaving" it - shouldn't be allowed, but lets assume it is)
4) First thread `free`s a `dev`
5) Seconds threads is no longer waits on a critical section - it is either successfully entered or got an error the the critical section has been destroyed
5) Second thread tries to access `dev`/read/etc. _after_ it was `free`d by `hid_close`

Reply:
This argument is invalid. It is normal for any kind of file operation on any kind of operating system that you cannot call any new file operations after calling close(). This is true for (almost) every operating system on the planet. This is the applications responsibility, just like on any other file system, as handles/pointers/tokens could be reused.

Reply:
then why is this (syncronisation) even needed?
so you can safely interrupt the read by hid_close? - That has to be a responsibility of a separate function

Reply:
> you cannot call any new file operations after calling close(). This is true for (almost) every operating system on the planet.

You're absolutely right, and let me extend this statement:
- you cannot call any file operation after entering the `close` function (even before it leaves);
- you cannot call `close` until you finish all file operations (leave each corresponding function);

Reply:
> This argument is invalid.

The scenario shouldn't be allowed. But if it happens - the argument is valid. We should distinct the two.

Reply:
Again, theoretically I agree with everything you state.

However, using this limitation which exists in basically _all_ file systems I have encountered as a reason to never have sane thread safety in a simple library is a stretch....